extends Node2D
class_name Sun

@export var sun_value := 25

## é˜³å…‰å­˜åœ¨æ—¶é—´
@export var exist_time:float = 10.0
var collected := false  # æ˜¯å¦å·²è¢«ç‚¹å‡»æ”¶é›†
## ç”Ÿäº§é˜³å…‰ç§»åŠ¨çš„tween
var spawn_sun_tween:Tween

func _ready() -> void:
	_sun_scale(sun_value)
	## å¯åŠ¨ä¸€ä¸ª10ç§’å®šæ—¶å™¨
	await get_tree().create_timer(exist_time, false).timeout

	# å¦‚æœè¿˜æ²¡è¢«ç‚¹å‡»æ”¶é›†ï¼Œè‡ªåŠ¨é”€æ¯
	if not collected and is_instance_valid(self):
		_start_fade_out()

func init_sun(curr_sun_value:int, pos:Vector2):
	sun_value = curr_sun_value
	position = pos

func _sun_scale(new_sun_value:int):
	var new_scale = new_sun_value/25.0
	scale = Vector2(new_scale,new_scale)


func _on_button_pressed() -> void:
	if spawn_sun_tween:
		spawn_sun_tween.kill()

	if collected:
		return  # é˜²æ­¢é‡å¤ç‚¹å‡»

	collected = true  # è®¾ç½®å·²è¢«æ”¶é›†
	var target_position = Vector2()
	SoundManager.play_other_SFX("points")
	if is_instance_valid(Global.main_game):
		if is_instance_valid(Global.main_game.marker_2d_sun_target):
			## å‡ºæˆ˜å¡æ§½åœ¨canvaslayerä¸­ï¼Œä½ç½®å’Œæ‘„åƒå¤´ä½ç½®æœ‰åç§»
			target_position = Global.main_game.marker_2d_sun_target.global_position + Global.main_game.camera_2d.global_position
			#print(Global.main_game.marker_2d_sun_target.get_canvas_layer_node().get_final_transform())
		else:
			target_position = Global.main_game.marker_2d_sun_target_default.global_position

		EventBus.push_event("add_sun_value", [sun_value])

	var tween:Tween = get_tree().create_tween()
	# å°†èŠ‚ç‚¹ä»å½“å‰ä½ç½®ç§»åŠ¨åˆ°(100, 200)ï¼Œè€—æ—¶0.5ç§’
	tween.tween_property(self, "global_position", target_position, 0.3).set_ease(Tween.EASE_OUT)
	$Button.queue_free()
	await tween.finished
	## åˆ°è¾¾ä½ç½®ï¼Œå˜é€æ˜
	tween = create_tween()
	tween.set_parallel()
	tween.tween_property(self, "modulate:a", 0, 0.5)
	await tween.finished
	queue_free()


func _start_fade_out() -> void:
	var tween = get_tree().create_tween()
	tween.tween_property(self, "modulate:a", 0.0, 1.0)  # ğŸŒ«ï¸ 1ç§’æ·¡å‡º
	tween.finished.connect(func():
		if not collected and is_instance_valid(self):
			self.queue_free()
	)

func on_sun_tween_finished():
	if Global.auto_collect_sun:
		_on_button_pressed()
